src/vector_calculus.f90

```diff
@@ -0,0 +1,346 @@
+    ! Interpolation for pressure field in z
+    call self%backend%tds_solve(p_sxy_z, pressure, z_interpl_p2v)
+    call self%backend%tds_solve(dpdz_sxy_z, pressure, z_stagder_p2v)
+
+    ! request fields from the allocator
+    p_sxy_y => self%backend%allocator%get_block(DIR_Y)
+    dpdz_sxy_y => self%backend%allocator%get_block(DIR_Y)
+
+    ! reorder data from z orientation to y orientation
+    call self%backend%reorder(p_sxy_y, p_sxy_z, RDR_Z2Y)
+    call self%backend%reorder(dpdz_sxy_y, dpdz_sxy_z, RDR_Z2Y)
+
+    call self%backend%allocator%release_block(p_sxy_z)
+    call self%backend%allocator%release_block(dpdz_sxy_z)
+
+    p_sx_y => self%backend%allocator%get_block(DIR_Y)
+    dpdy_sx_y => self%backend%allocator%get_block(DIR_Y)
+    dpdz_sx_y => self%backend%allocator%get_block(DIR_Y)
+
+    ! similar to the z direction, obtain derivatives in y.
+    call self%backend%tds_solve(p_sx_y, p_sxy_y, y_interpl_p2v)
+    call self%backend%tds_solve(dpdy_sx_y, p_sxy_y, y_stagder_p2v)
+    call self%backend%tds_solve(dpdz_sx_y, dpdz_sxy_y, y_interpl_p2v)
+
+    ! release memory
+    call self%backend%allocator%release_block(p_sxy_y)
+    call self%backend%allocator%release_block(dpdz_sxy_y)
+
+    ! just like in y direction, get some fields for the x derivatives.
+    p_sx_x => self%backend%allocator%get_block(DIR_X)
+    dpdy_sx_x => self%backend%allocator%get_block(DIR_X)
+    dpdz_sx_x => self%backend%allocator%get_block(DIR_X)
+
+    ! reorder from y to x
+    call self%backend%reorder(p_sx_x, p_sx_y, RDR_Y2X)
+    call self%backend%reorder(dpdy_sx_x, dpdy_sx_y, RDR_Y2X)
+    call self%backend%reorder(dpdz_sx_x, dpdz_sx_y, RDR_Y2X)
+
+    ! release all the y directional fields.
+    call self%backend%allocator%release_block(p_sx_y)
+    call self%backend%allocator%release_block(dpdy_sx_y)
+    call self%backend%allocator%release_block(dpdz_sx_y)
+
+    ! get the derivatives in x
+    call self%backend%tds_solve(dpdx, p_sx_x, x_stagder_p2v)
+    call self%backend%tds_solve(dpdy, dpdy_sx_x, x_interpl_p2v)
+    call self%backend%tds_solve(dpdz, dpdz_sx_x, x_interpl_p2v)
+
+    ! release temporary x fields
+    call self%backend%allocator%release_block(p_sx_x)
+    call self%backend%allocator%release_block(dpdy_sx_x)
+    call self%backend%allocator%release_block(dpdz_sx_x)
+
+  end subroutine gradient_p2v
+
+  subroutine laplacian(self, lapl_u, u, x_der2nd, y_der2nd, z_der2nd)
+    implicit none
+
+    class(vector_calculus_t) :: self
+    class(field_t), intent(inout) :: lapl_u
+    class(field_t), intent(in) :: u
+
+    class(tdsops_t), intent(in) :: x_der2nd, y_der2nd, z_der2nd
+
+    class(field_t), pointer :: u_y, d2u_y, u_z, d2u_z
+
+    ! d2u/dx2
+    call self%backend%tds_solve(lapl_u, u, x_der2nd)
+
+    ! y directional temporary fields
+    u_y => self%backend%allocator%get_block(DIR_Y)
+    d2u_y => self%backend%allocator%get_block(DIR_Y)
+
+    call self%backend%reorder(u_y, u, RDR_X2Y)
```

Comment:
is it worth checking that `u` is in `DIR_X` before doing that? or at least add a comment in the header to explain that it is expected

Reply:
`tds_solve` is doing a check based on directions so added a comment in the laplacian subroutine header to explain